autorest/authorization_storage.go

```
@@ -0,0 +1,298 @@
+type SharedKeyType string
+
+// see https://docs.microsoft.com/en-us/rest/api/storageservices/authorize-with-shared-key for details on the shared key types
+const (
+	// SharedKey is used to authorize against blobs, files and queues services.
+	SharedKey SharedKeyType = "sharedKey"
+
+	// SharedKeyForTable is used to authorize against the table service.
+	SharedKeyForTable SharedKeyType = "sharedKeyTable"
+
+	// SharedKeyLite is used to authorize against blobs, files and queues services.  It's provided for
+	// backwards compatibility with API versions before 2009-09-19.  Prefer SharedKey instead.
+	SharedKeyLite SharedKeyType = "sharedKeyLite"
+
+	// SharedKeyLiteForTable is used to authorize against the table service.  It's provided for
+	// backwards compatibility with older table API versions.  Prefer SharedKeyForTable instead.
+	SharedKeyLiteForTable SharedKeyType = "sharedKeyLiteTable"
+)
+
+const (
+	headerContentEncoding   = "Content-Encoding"
+	headerContentLength     = "Content-Length"
+	headerContentMD5        = "Content-MD5"
+	headerContentLanguage   = "Content-Language"
+	headerIfModifiedSince   = "If-Modified-Since"
+	headerIfMatch           = "If-Match"
+	headerIfNoneMatch       = "If-None-Match"
+	headerIfUnmodifiedSince = "If-Unmodified-Since"
+	headerDate              = "Date"
+	headerXMSDate           = "X-Ms-Date"
+	headerRange             = "Range"
+)
+
+const storageEmulatorAccountName = "devstoreaccount1"
+
+// SharedKeyAuthorizer implements an authorization for Shared Key
+// this can be used for interaction with Blob, File and Queue Storage Endpoints
+type SharedKeyAuthorizer struct {
+	accountName string
+	accountKey  []byte
+	keyType     SharedKeyType
+}
+
+// NewSharedKeyAuthorizer creates a SharedKeyAuthorizer using the provided credentials and shared key type.
+func NewSharedKeyAuthorizer(accountName, accountKey string, keyType SharedKeyType) (SharedKeyAuthorizer, error) {
```

I'd argue this should be:

```suggestion
func NewSharedKeyAuthorizer(accountName, accountKey string, keyType SharedKeyType) (*SharedKeyAuthorizer, error) {
```

since if there's an error there shouldn't be an authorizer returned?

---

I have it returning by value as the method set for this type is also by value (SharedKeyAuthorizer is immutable and copy semantics are ok).  And if error is non-nil the caller shouldn't be touching any other return values.

---

But do you trust the caller? :) Returning nil pointer here is safer here in my opinion.

---

We discussed this and will change to by-ref but not because it's safer (ignoring the error breaks the method's contract).  A `SharedKeyAuthorizer` is unique for a set of inputs so should be shared; in the future we could also add support for changing the account key so copy-semantics don't work there.  Also, the zero-value for `SharedKeyAuthorizer` is useless.